lib/authz/permissions.go

```diff
@@ -1355,3 +1355,27 @@ func UserFromContext(ctx context.Context) (IdentityGetter, error) {
 	}
 	return user, nil
 }
+
+// HasBuiltinRole checks if the identity is a builtin role with the matching
+// name.
+func HasBuiltinRole(authContext Context, name string) bool {
+	if _, ok := authContext.Identity.(BuiltinRole); !ok {
+		return false
+	}
+	if !authContext.Checker.HasRole(name) {
+		return false
+	}
+
+	return true
+}
+
+// IsLocalUser checks if the identity is a local user.
+func IsLocalUser(authContext Context) bool {
+	_, ok := authContext.Identity.(LocalUser)
+	return ok
+}
+
+// IsCurrentUser checks if the identity is a local user matching the given username
+func IsCurrentUser(authContext Context, username string) bool {
+	return IsLocalUser(authContext) && authContext.User.GetName() == username
+}
```

Comment:
suggestion: Pass by pointer instead of copying it

```suggestion
// HasBuiltinRole checks if the identity is a builtin role with the matching
// name.
func HasBuiltinRole(authContext *Context, name string) bool {
	if _, ok := authContext.Identity.(BuiltinRole); !ok {
		return false
	}
	if !authContext.Checker.HasRole(name) {
		return false
	}

	return true
}

// IsLocalUser checks if the identity is a local user.
func IsLocalUser(authContext *Context) bool {
	_, ok := authContext.Identity.(LocalUser)
	return ok
}

// IsCurrentUser checks if the identity is a local user matching the given username
func IsCurrentUser(authContext *Context, username string) bool {
	return IsLocalUser(authContext) && authContext.User.GetName() == username
}
```

Reply:
This function used to take this argument by value, and I can change them to take pointers, I don't think this is the best idea. Pointers can save a few CPU cycles, but then you need to deal with potential `nil` pointer and for loop scope pointer issue.